---
title: "Map scRNA-seq data to maize AGPv5"
author: "Jinliang Yang and subhash"
date: "08-16-2024"
output: NULL
---


# 2.4 Data analysis of scRNA-seq of Maize
- text below is copied from the Li et al., 2022 paper

To generate the gene expression matrix of single cells, the raw reads were aligned to the B73 reference genome, and gene expression was quantified using the `Cell Ranger pipeline (version 3.1, 10X Genomics)`. 
The B73 reference genome and annotation files were downloaded from Ensembl (ftp://ftp.ensemblgenomes.org/ pub/plants/release-42). The gene annotation (gtf) file was filtered by the `cellranger mkgtf` function with the `--attribute = gene_bio type: protein_coding` argument. 
The `cellranger mkref` function of Cell Ranger was used to build a reference. The `cellranger count` function was performed to generate a raw count matrix.

# Cell Ranger installation



here is the [tutorials](https://www.10xgenomics.com/support/software/cell-ranger/latest/tutorials)

```{bash}
cd ~/bin/
wget -O cellranger-8.0.1.tar.gz "https://cf.10xgenomics.com/releases/cell-exp/cellranger-8.0.1.tar.gz?Expires=1723872327&Key-Pair-Id=APKAI7S6A5RYOXBWRPDA&Signature=QMT75~0APN1ylbKNNePQ6cVDlqqkgCAwzdCcX5lsc5v5QBJsMw48x4UHPYJcu8TYRJO7JF1EUkZnrQgNDae8qsUY~~wOLxckFia4gw1chgDOfW96J9ylhpNN0lOaXUy8deeiogn0NFoFMPPX67V-VjZWyJ6vo5hlySaT58sgewghGZPePoJIWXR~o78f1I7AIqOg8gQaP7AfDXROWxcd3nZTswklG1BAmRDMYQnB0zEGYYgvTxjEYDHUpjk3cuE5o7Uz0bldY9Cvw4TF~JdaXEPvH2HsTBBM~7-gZ~AlM4AwWwIM48wxAaHLR8BAxVztpSH~Kv9ftk4DJEk8rAKSjA__"

tar -xvzf cellranger-8.0.1.tar.gz

#export your path to .bashrc or .bash_profile

cellranger testrun --id=check_install

# if install successfully, you should get the below info:
```

Waiting 6 seconds for UI to do final refresh.
Pipestance completed successfully!

2024-08-16 13:49:40 Shutting down.
Saving pipestance info to "check_install/check_install.mri.tgz"

# Prepare the gene annotation file

https://www.maizegdb.org/genome/assembly/Zm-B73-REFERENCE-NAM-5.0

```{bash}
# get the reference genome and annotations
wget https://download.maizegdb.org/Zm-B73-REFERENCE-NAM-5.0/Zm-B73-REFERENCE-NAM-5.0.fa.gz
gzip -d Zm-B73-REFERENCE-NAM-5.0.fa.gz

wget https://download.maizegdb.org/Zm-B73-REFERENCE-NAM-5.0/Zm-B73-REFERENCE-NAM-5.0_Zm00001eb.1.gff3.gz
gzip -d Zm-B73-REFERENCE-NAM-5.0_Zm00001eb.1.gff3.gz
```

# Maize GFF file

```{bash}
cellranger mkgtf -h
cellranger mkgtf Zm-B73-REFERENCE-NAM-5.0_Zm00001eb.1.gff3 Zm-B73-REFERENCE-NAM-5.0_gene.gtf --attribute = gene
```

## preparing refrence genome

```{bash}
#converting downloaded .gff3 to .gtf using gffread
gffread Zm-B73-REFERENCE-NAM-5.0_Zm00001eb.1.gff3 \
-T -o Zm-B73-REFERENCE-NAM-5.0_Zm00001eb.1.gtf

#creating the index for refrence using cellranger mkgtf tool
cellranger mkgtf Zm-B73-REFERENCE-NAM-5.0_Zm00001eb.1.gtf \
Zm-B73-REFERENCE-NAM-5.0_Zm00001eb.1_filtered.gtf \
--attribute=gene_biotype:protein_coding

```

### alignment to refrence genome and generatig feature count matrix using cell ranger
In our study we have 2 experimental conditions with_nitrate and without_nitrate

```{bash}
#!/bin/sh
#SBATCH --ntasks-per-node=8
#SBATCH --nodes=1
#SBATCH --mem=64G
#SBATCH --time=80:00:00
#SBATCH --job-name=CR_count_2
#SBATCH --mail-user=smahamkalivenkatas2@huskers.unl.edu
#SBATCH --mail-type=ALL
#SBATCH --error=/work/jyanglab/subhash/3.single_cell_RNA_Seq/2.sc_RNASeq/3.CR_count/2.CR_count_2/0.scripts/2_CR.err
#SBATCH --output=/work/jyanglab/subhash/3.single_cell_RNA_Seq/2.sc_RNASeq/3.CR_count/2.CR_count_2/0.scripts/2_CR.out

cd /work/jyanglab/subhash/3.single_cell_RNA_Seq/2.sc_RNASeq/3.CR_count/2.CR_count_2/

cellranger count \
  --id=2_without_nitrate \
  --transcriptome=/work/jyanglab/subhash/3.single_cell_RNA_Seq/2.sc_RNASeq/2.refrence/Zm_B73 \
  --fastqs=/work/jyanglab/subhash/3.single_cell_RNA_Seq/2.sc_RNASeq/0.2.without_nitrate/ \
  --sample=SRR15686125,SRR15686126,SRR15686127,SRR15686128,SRR15686129,SRR15686130,SRR15686131,SRR15686132 \
  --localcores=8 \
  --localmem=64 \
  --create-bam=true
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir=normalizePath('/Users/subhashmahamkali/Documents/Root-cell-type-specific-genes-for-nitrate-response/')) 
```






#testing seurat package pipeline workflow

```{r}
#load libraries
library(Seurat)
library(ggplot2)
library(tidyverse)
library(dplyr)
library(gridExtra)
```



```{r}
# reading both with N and W.out N data 
dirs <- list.dirs(path = '/Users/subhashmahamkali/Downloads/sc_RNA_seq', recursive = F, full.names = F)
for(x in dirs){ 
  name <- gsub('_filtered_feature_bc_matrix','', x)
  
  cts <- ReadMtx(mtx = paste0('/Users/subhashmahamkali/Downloads/sc_RNA_seq/',x,'/matrix.mtx.gz'),
         features = paste0('/Users/subhashmahamkali/Downloads/sc_RNA_seq/',x,'/features.tsv.gz'),
         cells = paste0('/Users/subhashmahamkali/Downloads/sc_RNA_seq/',x,'/barcodes.tsv.gz'))
  
  # create seurat objects
  assign(name, CreateSeuratObject(counts = cts, min.cells = 3, min.features = 200))
}

N
#An object of class Seurat 
#24288 features across 2842 samples within 1 assay 
#Active assay: RNA (24288 features, 0 variable features)
#1 layer present: counts

WN
#An object of class Seurat 
#24225 features across 2777 samples within 1 assay 
#Active assay: RNA (24225 features, 0 variable features)
#1 layer present: counts

#grep("^mt-", rownames(N@assays$RNA@data), value = T) -#need to filterout mitochondria and chloroplast genes - not done yet.

features_N <- rownames(N)
features_WN <- rownames(WN)
common_features <- intersect(features_N, features_WN) #finding the common genes between both datasets and then filtering with these genes alone.
N <- subset(N, features = common_features)
WN <- subset(WN, features = common_features)

# merge datasets
merged_seurat <- merge(N, y = c(WN),
      add.cell.ids = c("N", "WN"),
      project = 'NI')
merged_seurat
#An object of class Seurat 
#23435 features across 5619 samples within 1 assay 
#Active assay: RNA (23435 features, 0 variable features)
#2 layers present: counts.1, counts.2

```

```{r}
# QC & filtering -----------------------
View(merged_seurat@meta.data)
# create a sample column
merged_seurat$sample <- rownames(merged_seurat@meta.data)
# split sample column
merged_seurat@meta.data <- separate(merged_seurat@meta.data, col = 'sample', into = c('Treatment', 'Barcode'), 
         sep = '_')

#merged_seurat$mitoPercent <- PercentageFeatureSet(merged_seurat, pattern='^MT-') #not done yet, need to filt
#v = VlnPlot(merged_seurat, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
v1 = VlnPlot(merged_seurat, features = c("nFeature_RNA"))+ NoLegend()
v2 = VlnPlot(merged_seurat, features = c("nCount_RNA"))+ NoLegend()

# explore QC
# filtering
merged_seurat_filtered <- subset(merged_seurat, subset = nFeature_RNA > 400 & nFeature_RNA < 5000)
merged_seurat_filtered
#An object of class Seurat 
#23435 features across 5093 samples within 1 assay 
#Active assay: RNA (23435 features, 0 variable features)
#2 layers present: counts.1, counts.2

v3 = VlnPlot(merged_seurat_filtered, features = c("nFeature_RNA"))+ NoLegend()
v4 = VlnPlot(merged_seurat_filtered, features = c("nCount_RNA"))+ NoLegend()
vgrid <- grid.arrange(v1,v2,v3,v4, ncol = 2, nrow = 2)
ggsave("graphs/SC_QC.pdf", vgrid, width = 15, height = 10)

range(merged_seurat_filtered$nFeature_RNA)
#402 4995 -range of unique number of genes per cell

scatp <- FeatureScatter(merged_seurat_filtered, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") +
  geom_smooth(method = 'lm')
ggsave("graphs/scaterplot.pdf", scatp, width = 15, height = 10)

# perform standard workflow steps to figure out if we see any batch effects --------
merged_seurat_filtered <- NormalizeData(object = merged_seurat_filtered)
merged_seurat_filtered <- FindVariableFeatures(object = merged_seurat_filtered)

top10 <- head(VariableFeatures(merged_seurat_filtered), 10)
#"Zm00001eb173060" "Zm00001eb090910" "Zm00001eb065710" "Zm00001eb373190" "Zm00001eb335210" "Zm00001eb249210" "Zm00001eb026630" "Zm00001eb123800" "Zm00001eb372230" "Zm00001eb266840"
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(merged_seurat_filtered)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
ggsave("graphs/variable_features.pdf", plot2, width = 15, height = 10) #top 10 labeled and 2000(red color) variable genes

merged_seurat_filtered <- ScaleData(object = merged_seurat_filtered)
merged_seurat_filtered <- RunPCA(object = merged_seurat_filtered, npcs = 50)
eplot <- ElbowPlot(merged_seurat_filtered, ndims = 50) + geom_vline(xintercept = 27, color = "red", linetype = "dashed") #27 pcs according to the paper
ggsave("graphs/elbow_plot.pdf", eplot, width = 15, height = 10)

#LabelPoints(plot = plot1, points = top10, repel = TRUE)

merged_seurat_filtered <- FindNeighbors(object = merged_seurat_filtered, dims = 1:27)
merged_seurat_filtered <- FindClusters(object = merged_seurat_filtered)
View(merged_seurat_filtered@meta.data)

merged_seurat_filtered <- FindClusters(object = merged_seurat_filtered, resolution = c(0.1,0.2,0.3, 0.5, 0.7, 10))
DimPlot(merged_seurat_filtered, group.by = "RNA_snn_res.0.3", label = TRUE)

#merged_seurat_filtered <- FindClusters(object = merged_seurat_filtered, resolution = c(0.1,0.2,0.3, 0.5, 0.7, 10))
#DimPlot(merged_seurat_filtered, group.by = "RNA_snn_res.0.3", label = TRUE)
#Idents(pr1s) <- "RNA_snn_res.0.3"

merged_seurat_filtered <- RunUMAP(object = merged_seurat_filtered, dims = 1:27) 

# plot
p <- DimPlot(merged_seurat_filtered, reduction = 'umap', group.by = 'Treatment',
        cols = c('red','green'))+ ggtitle("Before integration and batch correction")


# perform integration to correct for batch effects ------
obj.list <- SplitObject(merged_seurat_filtered, split.by = 'Treatment')
for(i in 1:length(obj.list)){
  obj.list[[i]] <- NormalizeData(object = obj.list[[i]])
  obj.list[[i]] <- FindVariableFeatures(object = obj.list[[i]])
  #obj.list[[i]] <- RunPCA(obj.list[[i]], npcs = 50)
}
# select integration features
features <- SelectIntegrationFeatures(object.list = obj.list)
# find integration anchors (CCA)
anchors <- FindIntegrationAnchors(object.list = obj.list,
                       anchor.features = features,)
                       #reduction = "rpca")
# integrate data
seurat.integrated <- IntegrateData(anchorset = anchors)
# Scale data, run PCA and UMAP and visualize integrated data
seurat.integrated <- ScaleData(object = seurat.integrated)
seurat.integrated <- RunPCA(object = seurat.integrated, npcs = 30)
ElbowPlot(merged_seurat_filtered, ndims = 30)
seurat.integrated <- RunUMAP(object = seurat.integrated, dims = 1:27)

p2 <- DimPlot(seurat.integrated, reduction = 'umap', group.by = 'Treatment',
              cols = c('red','green')) + ggtitle("After integration and batch correction")

Tumap <- grid.arrange(p, p2, ncol = 2)
ggsave("graphs/T2UMAP.pdf", Tumap, width = 15, height = 10)
```

```{r}
#Clustering ------------
#cluster similar cells
#cells with similar expression parents

seurat.integrated <- FindNeighbors(seurat.integrated, dims = 1:27) #most variation is captured in our dataset
# understanding resolution
#in this step we want cells to be asigned to the clusters. for the res here defines granularity and resoultion of the clusters
seurat.integrated <- FindClusters(seurat.integrated, resolution = c(0.1,0.2,0.3, 0.5, 0.7, 1)) #low number - less clusters and high numnber - more clusters -trail and error.
view(seurat.integrated@meta.data)
DimPlot(seurat.integrated, reduction = "umap", group.by = "RNA_snn_res.0.3", label = TRUE)

# setting identity of clusters
Idents(seurat.integrated)
Idents(seurat.integrated) <- "RNA_snn_res.0.3"

# non-linear dimensionality reduction
seurat.integrated <- RunUMAP(seurat.integrated, dims = 1:27)
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
DimPlot(seurat.integrated, reduction = "umap")
levels(seurat.integrated)
desired_order <- c("0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10")
seurat.integrated@active.ident <- factor(seurat.integrated@active.ident, levels = desired_order)
print(levels(seurat.integrated))
umap1 <- DimPlot(seurat.integrated, reduction = "umap", label = TRUE)

markers <- FindAllMarkers(seurat.integrated, test.use = "bimod", logfc.threshold = 1)
top_markers_all <- markers %>% group_by(cluster) %>% top_n(n = 11, wt = avg_log2FC)
top_markers_all

#gene_search <- markers %>% filter(gene == "Zm00001eb137050")
#gene_search <- markers %>% filter(gene %in% filtered_data1$v5.Gene.Model.ID) #filt
#gene_search
#Top <- top_markers_all %>% filter(gene %in% filtered_data1$v5.Gene.Model.ID)

zero <- FeaturePlot(seurat.integrated, c("Zm00001eb175860"), min.cutoff = 'q10') + ggtitle("cluster0(stele)") 
one <- FeaturePlot(seurat.integrated, c("Zm00001eb429560"), min.cutoff = 'q10') + ggtitle("cluster1(cortex)") 
two <- FeaturePlot(seurat.integrated, c("Zm00001eb152160"), min.cutoff = 'q10') + ggtitle("cluster2(CC)")
#three <- FeaturePlot(seurat.integrated, c("Zm00001eb152160"), min.cutoff = 'q10') + ggtitle("Zm00001eb152160 - cluster2(CC)")
four <- FeaturePlot(seurat.integrated, c("Zm00001eb091170"), min.cutoff = 'q10') + ggtitle("luster4(root cap)")
five <- FeaturePlot(seurat.integrated, c("Zm00001eb199490"), min.cutoff = 'q10') + ggtitle("luster5(exodermis)")
six <- FeaturePlot(seurat.integrated, c("Zm00001eb137050"), min.cutoff = 'q10') + ggtitle("cluster6(epidermis)")
seven <- FeaturePlot(seurat.integrated, c("Zm00001eb173060"), min.cutoff = 'q10') + ggtitle("cluster7(protophloem_sieve)")
eight <- FeaturePlot(seurat.integrated, c("Zm00001eb187430"), min.cutoff = 'q10') + ggtitle("cluster8(pericycle)")
nine <- FeaturePlot(seurat.integrated, c("Zm00001eb025100"), min.cutoff = 'q10') + ggtitle("cluster9(root hair)")
ten <- FeaturePlot(seurat.integrated, c("Zm00001eb080730"), min.cutoff = 'q10') + ggtitle("cluster10(endodermis)")


pgrid <- grid.arrange(zero, one, two, four, five, six, seven, eight, nine, ten,  ncol = 5 )
ggsave("graphs/SC_feature_plots.pdf", pgrid, width = 15, height = 10)

#0-Zm00001eb175860
#1-Zm00001eb429560
#2-Zm00001eb152160
#4-Zm00001eb091170
#5-Zm00001eb199490
#6-Zm00001eb137050
#7-Zm00001eb173060 
#8-Zm00001eb187430
#9-Zm00001eb025100
#10-Zm00001eb080730
#GLR1: glutamic acid and lysine-rich1  Zm00001eb119130
```

```{r}
DefaultAssay(seurat.integrated)
DefaultAssay(seurat.integrated) <- "RNA"
seurat.integrated <- JoinLayers(seurat.integrated)
DefaultAssay(seurat.integrated) <- "integrated"
new.cluster.ids <- c("0- Stele",
                     "1- cortex",
                     "2- CC",
                     "3- mz",
                     "4- root cap",
                     "5- exodermis",
                     "6- epidermis",
                     "7- protophloem_sive",
                     "8- pericycle",
                     "9- root hair",
                     "10- endodermis")
seu.labeled <- seurat.integrated
names(new.cluster.ids) <- levels(seu.labeled)
seu.labeled <- RenameIdents(seu.labeled, new.cluster.ids)
umap2 <- DimPlot(seu.labeled, label = T)

# Get the cluster identities of each cell
cluster_counts <- table(Idents(seurat.integrated))
# Display the number of cells in each cluster
print(cluster_counts)
new.cluster.ids <- paste0(new.cluster.ids, " (", cluster_counts, ")")
names(new.cluster.ids) <- levels(seurat.integrated)
seu.labeled <- RenameIdents(seurat.integrated, new.cluster.ids)
# Plot the UMAP with the updated labels
umap3 <- DimPlot(seu.labeled, label = TRUE, label.size = 4)
ggsave("graphs/clustered_umap.pdf", umap3, width = 15, height = 10)



#glr1(glutamic acid- and lysine-rich1): Zm00001eb119130 expression in cell types
glr <- FeaturePlot(seurat.integrated, c("Zm00001eb119130"), min.cutoff = 'q10') + ggtitle("glr1(glutamic acid- and lysine-rich1)") 
ggsave("graphs/glr.png", glr, width = 15, height = 10)


#0-
#1-
#2-
#4-
#5-
#6-
#7- 
#8-
#9-
#10-

cluster_specific_genes <- c("Zm00001eb175860", "Zm00001eb429560","Zm00001eb152160", "Zm00001eb091170", "Zm00001eb199490", "Zm00001eb137050", "Cd209a", "Clec4a4", "Zm00001eb173060", "Zm00001eb187430", "Zm00001eb025100", "Zm00001eb080730")
# can use a RColorBrewer pallete, I chose RdBu here
knownmarker <- DotPlot(seurat.integrated, features = rev(cluster_specific_genes), cols = "RdBu") + theme(axis.text.x = element_text( angle = 90, vjust = 0.5, hjust = 1))
ggsave("graphs/known_markers.png", knownmarker, width = 15, height = 10)
```


```{r}
#GLR gene expression 

#{ZmGLR3.4
#Zm00001eb225110
#Zm00001eb225150
#Zm00001eb225130}

ZmGLR3.4_1 <- FeaturePlot(seurat.integrated, c("Zm00001eb225110"), min.cutoff = 'q10') + ggtitle("ZmGLR3.4_1") 
ZmGLR3.4_2 <- FeaturePlot(seurat.integrated, c("Zm00001eb225150"), min.cutoff = 'q10') + ggtitle("ZmGLR3.4_2") 
ZmGLR3.4_3 <- FeaturePlot(seurat.integrated, c("Zm00001eb225130"), min.cutoff = 'q10') + ggtitle("ZmGLR3.4_3") 
ZmGLR3.4 <- grid.arrange(ZmGLR3.4_1, ZmGLR3.4_2, ZmGLR3.4_3, ncol = 3 )
ggsave("graphs/ZmGLR3.4.pdf", ZmGLR3.4, width = 15, height = 5)



#{ZmGLR3.5
#Zm00001eb106130
#Zm00001eb298150}
ZmGLR3.5_1 <- FeaturePlot(seurat.integrated, c("Zm00001eb106130"), min.cutoff = 'q10') + ggtitle("ZmGLR3.5_1") 
ZmGLR3.5_2 <- FeaturePlot(seurat.integrated, c("Zm00001eb298150"), min.cutoff = 'q10') + ggtitle("ZmGLR3.5_2") 
ZmGLR3.5 <- grid.arrange(ZmGLR3.5_1, ZmGLR3.5_2, ncol = 3 )
ggsave("graphs/ZmGLR3.5.pdf", ZmGLR3.5, width = 15, height = 5)


#{ZmGLR3.3
#Zm00001eb072720
#Zm00001eb230060
#Zm00001eb209370}

#{ZmGLR3.7
#Zm00001eb192720
#Zm00001eb257400}


#Zm00001eb099820: ZmGLR2.7c
#Zm00001eb313570: ZmGLR2.7b

#Zm00001eb313550: ZmGLR2









```

