---
title: "Map scRNA-seq data to maize AGPv5"
author: "Jinliang Yang and subhash"
date: "08-16-2024"
output: NULL
---


# 2.4 Data analysis of scRNA-seq of Maize
- text below is copied from the Li et al., 2022 paper

To generate the gene expression matrix of single cells, the raw reads were aligned to the B73 reference genome, and gene expression was quantified using the `Cell Ranger pipeline (version 3.1, 10X Genomics)`. 
The B73 reference genome and annotation files were downloaded from Ensembl (ftp://ftp.ensemblgenomes.org/ pub/plants/release-42). The gene annotation (gtf) file was filtered by the `cellranger mkgtf` function with the `--attribute = gene_bio type: protein_coding` argument. 
The `cellranger mkref` function of Cell Ranger was used to build a reference. The `cellranger count` function was performed to generate a raw count matrix.

# Cell Ranger installation

here is the [tutorials](https://www.10xgenomics.com/support/software/cell-ranger/latest/tutorials)

```{bash}
cd ~/bin/
wget -O cellranger-8.0.1.tar.gz "https://cf.10xgenomics.com/releases/cell-exp/cellranger-8.0.1.tar.gz?Expires=1723872327&Key-Pair-Id=APKAI7S6A5RYOXBWRPDA&Signature=QMT75~0APN1ylbKNNePQ6cVDlqqkgCAwzdCcX5lsc5v5QBJsMw48x4UHPYJcu8TYRJO7JF1EUkZnrQgNDae8qsUY~~wOLxckFia4gw1chgDOfW96J9ylhpNN0lOaXUy8deeiogn0NFoFMPPX67V-VjZWyJ6vo5hlySaT58sgewghGZPePoJIWXR~o78f1I7AIqOg8gQaP7AfDXROWxcd3nZTswklG1BAmRDMYQnB0zEGYYgvTxjEYDHUpjk3cuE5o7Uz0bldY9Cvw4TF~JdaXEPvH2HsTBBM~7-gZ~AlM4AwWwIM48wxAaHLR8BAxVztpSH~Kv9ftk4DJEk8rAKSjA__"

tar -xvzf cellranger-8.0.1.tar.gz

#export your path to .bashrc or .bash_profile

cellranger testrun --id=check_install

# if install successfully, you should get the below info:
```

Waiting 6 seconds for UI to do final refresh.
Pipestance completed successfully!

2024-08-16 13:49:40 Shutting down.
Saving pipestance info to "check_install/check_install.mri.tgz"

# Prepare the gene annotation file

https://www.maizegdb.org/genome/assembly/Zm-B73-REFERENCE-NAM-5.0

```{bash}
# get the reference genome and annotations
wget https://download.maizegdb.org/Zm-B73-REFERENCE-NAM-5.0/Zm-B73-REFERENCE-NAM-5.0.fa.gz
gzip -d Zm-B73-REFERENCE-NAM-5.0.fa.gz

wget https://download.maizegdb.org/Zm-B73-REFERENCE-NAM-5.0/Zm-B73-REFERENCE-NAM-5.0_Zm00001eb.1.gff3.gz
gzip -d Zm-B73-REFERENCE-NAM-5.0_Zm00001eb.1.gff3.gz
```

# Maize GFF file

```{bash}
cellranger mkgtf -h
cellranger mkgtf Zm-B73-REFERENCE-NAM-5.0_Zm00001eb.1.gff3 Zm-B73-REFERENCE-NAM-5.0_gene.gtf --attribute = gene
```

## preparing refrence genome

```{bash}
#converting downloaded .gff3 to .gtf using gffread
gffread Zm-B73-REFERENCE-NAM-5.0_Zm00001eb.1.gff3 \
-T -o Zm-B73-REFERENCE-NAM-5.0_Zm00001eb.1.gtf

#creating the index for refrence using cellranger mkgtf tool
cellranger mkgtf Zm-B73-REFERENCE-NAM-5.0_Zm00001eb.1.gtf \
Zm-B73-REFERENCE-NAM-5.0_Zm00001eb.1_filtered.gtf \
--attribute=gene_biotype:protein_coding

```

### alignment to refrence genome and generatig feature count matrix using cell ranger
In our study we have 2 experimental conditions with_nitrate and without_nitrate

```{bash}
#!/bin/sh
#SBATCH --ntasks-per-node=8
#SBATCH --nodes=1
#SBATCH --mem=64G
#SBATCH --time=80:00:00
#SBATCH --job-name=CR_count_2
#SBATCH --mail-user=smahamkalivenkatas2@huskers.unl.edu
#SBATCH --mail-type=ALL
#SBATCH --error=/work/jyanglab/subhash/3.single_cell_RNA_Seq/2.sc_RNASeq/3.CR_count/2.CR_count_2/0.scripts/2_CR.err
#SBATCH --output=/work/jyanglab/subhash/3.single_cell_RNA_Seq/2.sc_RNASeq/3.CR_count/2.CR_count_2/0.scripts/2_CR.out

cd /work/jyanglab/subhash/3.single_cell_RNA_Seq/2.sc_RNASeq/3.CR_count/2.CR_count_2/

cellranger count \
  --id=2_without_nitrate \
  --transcriptome=/work/jyanglab/subhash/3.single_cell_RNA_Seq/2.sc_RNASeq/2.refrence/Zm_B73 \
  --fastqs=/work/jyanglab/subhash/3.single_cell_RNA_Seq/2.sc_RNASeq/0.2.without_nitrate/ \
  --sample=SRR15686125,SRR15686126,SRR15686127,SRR15686128,SRR15686129,SRR15686130,SRR15686131,SRR15686132 \
  --localcores=8 \
  --localmem=64 \
  --create-bam=true
```









```{r}
library(Seurat)
library(tidyverse)

# Load the NSCLC dataset
#nsclc.sparse.m <- Read10X_h5(filename = 'filtered_feature_bc_matrix.h5')
nsclc.sparse.m <- Read10X(data.dir = 'filtered_feature_bc_matrix/')
str(nsclc.sparse.m)


pr1 <- CreateSeuratObject(counts = nsclc.sparse.m, project = "pr1", min.cells = 3, min.features = 200)
class(pr1) #seurat object
pr1
#An object of class Seurat 
#24225 features across 2777 samples within 1 assay 
#Active assay: RNA (24225 features, 0 variable features)

colnames(pr1[]) #barcodes
rownames(pr1[]) #genes

view(pr1)
view(pr1@meta.data)
#barcode,#projectname,#ncount_RNA,nfeature_RNA

#save seurat object
saveRDS(pr1, file="/Users/subhashmahamkali/Desktop/pr1.RDS")
#pr1 <- readRDS("/Users/subhashmahamkali/Desktop/pr1.RDS")
#pr1


#vid-3
merged <- merge(pr1, y = c(pr2),
                add.cell.ids = ls()[1:2],
                project = 'merged')
  

#quality control and selecting cells fro further analysis
#data normalization

range(pr1$nFeature_RNA)
range(pr1$nCount_RNA)

#store mitovhondrial percentage in object meta data (percentahe pf mitochroa genes for each cell.)


pr1 <- PercentageFeatureSet(pr1, pattern = "^MT-", col.name = "percent.mt")

view(pr1@meta.data)
range(pr1$percent.mt)
str(pr1)

library(rtracklayer)
library(Seurat)


# Install and load the rtracklayer package
if (!requireNamespace("BiocManager", quietly = TRUE)) {
    install.packages("BiocManager")
}

BiocManager::install("rtracklayer")
library(rtracklayer)

mito_genes <- gtf[grepl("mitochondrial", mcols(gtf)$gene_biotype, ignore.case = TRUE)]
mito_gene_ids <- mcols(mito_genes)$gene_id
pr1[["percent.mt"]] <- PercentageFeatureSet(pr1, features = mito_gene_ids)

view(pr1@meta.data)
```

